File: inmet/satorbit.py
# Copyright (C) 2018  István Bozsó
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

import numpy as np
from os.path import isfile
import pickle as pk

from gnuplot import Gnuplot, linedef
from inmet.utils import get_par
import inmet.inmet_aux as ina

__all__ = ("Satorbit")

class Satorbit(object):
    def __init__(self, path, mode):
        
        if mode == "fit_file":
            self.read_fit(path)
        elif mode == "doris" or mode == "gamma":
            self.read_orbits(path, mode)
    
    def read_orbits(self, path, preproc):
    
        if not isfile(path):
            raise IOError("{} is not a file.".format(path))
    
        with open(path, "r") as f:
            lines = f.readlines()
        
        if preproc == "doris":
            datanum = [(ii, line) for ii, line in enumerate(lines)
                        if line.startswith("NUMBER_OF_DATAPOINTS:")]
        
            if len(datanum) != 1:
                raise ValueError("More than one or none of the lines contain the "
                                 "number of datapoints.")
        
            idx = datanum[0][0]
            datanum = int(datanum[0][1].split(":")[1])
            
            data = np.fromstring("".join(lines[idx + 1:idx + datanum + 1]),
                                 count=datanum * 4, dtype=np.double, sep=" ")\
                                 .reshape((datanum, 4))
    
            time = data[:,0]
            coords = data[:,1:]
        
        elif preproc == "gamma":
            datanum = [(ii, line) for ii, line in enumerate(lines)
                       if line.startswith("number_of_state_vectors:")]
    
            if len(datanum) != 1:
                raise ValueError("More than one or none of the lines contains the "
                                 "number of datapoints.")
            
            idx = datanum[0][0]
            datanum = int(datanum[0][1].split(":")[1])
            
            t_first = float(lines[idx + 1].split(":")[1].split()[0])
            t_step  = float(lines[idx + 2].split(":")[1].split()[0])
            
            time = np.arange(t_first, t_first + 7 * t_step, t_step)
            
            coords = \
            np.asarray([str2orbit(get_par(f"state_vector_position_{ii+1}",
                        lines)) for ii in range(data_num)])
            
        else:
            raise ValueError("preproc should be either \"doris\" or \"gamma\" "
                             "not {}".format(preproc))
        
        self.time = time
        self.coords = coords
        self.datanum = datanum


    def read_fit(self, fit_file):
        
        with open(fit_file, "r") as f:
            poly = {line.split(":")[0]: line.split(':')[1].strip()
                    for line in f if ":" in line}
        
        self.centered = int(poly["centered"])
        self.deg = int(poly["deg"])
        
        if self.centered:
            self.mean_coords = np.genfromtxt(poly["mean_coords"].split(),
                                             dtype=np.double)
            self.t_mean      = float(poly["mean_time"])
        else:
            self.mean_coords = [0.0, 0.0, 0.0]
            self.t_mean      = 0.0
        
        self.t_start = float(poly["t_start"])
        self.t_stop  = float(poly["t_stop"])
        
        self.coeffs = np.genfromtxt(poly["coefficients"].split(),
                                    dtype=np.double).reshape(3, deg + 1)

    
    def fit_orbit(self, centered=True, deg=3):
        
        time, coords = self.time, self.coords
        
        self.centered = centered
        self.deg = deg
        
        self.t_start = np.min(time)
        self.t_stop  = np.max(time)
        
        if centered:
            mean_t = np.mean(time)
            mean_coords = np.mean(coords, axis=0)
    
            time -= mean_t
            to_fit = coords - mean_coords
        else:
            to_fit = coords
            cent = "centered:\t0\n"
        
        design = np.vander(time, deg + 1)
        
        # coeffs[0]: polynom coeffcients are in the columns
        # coeffs[1]: residuals
        # coeffs[2]: rank of design matrix
        # coeffs[3]: singular values of design matrix
        self.coeffs = np.linalg.lstsq(design, to_fit, rcond=None)
        
    def save_fit(self, savefile):
        
        time, coeffs = self.time, self.coeffs
        
        if self.centered:
            cent = "centered:\t1\n"
        else:
            cent = "centered:\t0\n"
            
        with open(savefile, "w") as f:
            f.write(cent)
        
            f.write("Start and stop times.\n")
            f.write("t_start:\t{}\n".format(self.t_start))
            f.write("t_stop:\t{}\n".format(self.t_stop))
            f.write("Degree of fitted polynom.\n")
            f.write("deg:\t{}\n".format(self.deg))
            
            f.write("(x, y, z) residuals: ({})\n"
                    .format(", ".join(str(elem) for elem in coeffs[1].astype(str))))
    
            f.write("\nCoeffcients are written in a single line from highest to\
                     \nlowest power. First for x than for y and finally for z.\n\n")
            f.write("coefficients:\t{}\n"
                    .format(" ".join(str(elem)
                                for elem in coeffs[0].T.reshape(-1).astype(str))))
            
            if self.centered:
                f.write("mean_time:\t{}\n".format(mean_t))
                f.write("mean_coords:\t{}\n"
                        .format(" ".join(str(coord) for coord in self.mean_coords)))
        

    def azi_inc(fit_file, coords, is_lonlat=True, max_iter=1000):
        
        return ina.azi_inc(self.t_start, self.t_stop, self.t_mean,
                           self.mean_coords, self.centered, self.coeffs,
                           self.deg, max_iter, is_lonlat, coords)

    def plot_orbit(self, plotfile, nsamp=100):
        
        coeffs = self.coeffs[0].T
        coords = self.coords / 1e3
        
        t = self.time
        time = np.linspace(self.t_start, self.t_stop, nsamp)
        
        if self.centered:
            time_cent = time - self.t_mean
            
            poly = np.asarray([np.polyval(coeffs[ii,:], time_cent)
                               + mean_coords[ii] for ii in range(3)]).T
            
        else:
            poly = np.asarray([np.polyval(coeffs[ii,:], time)
                               for ii in range(3)]).T
        
        poly /= 1e3
        
        gpt = Gnuplot()
        
        gpt.output(plotfile, term="pngcairo", fontsize=8)
        
        gpt.axis_format("x", "")
        gpt.margins(bmargin=3.5)
        gpt("set tics font ',8'")
        
        gpt.multiplot(3, portrait=True, title="Fit of orbital vectors - "
                      "degree of polynom: {}".format(self.deg))
        
        ylabels = ("X [km]", "Y [km]", "Z [km]")
        
        
        for ii in range(3):
            
            gpt.ylabel(ylabels[ii])
            
            points = gpt.data(t, coords[:,ii], title="Orbital coordinates",
                              vith=linedef(point_type="empty_circle"))
            
            fit = gpt.data(time, poly[:,ii], title="Fitted polynom",
                           vith="lines")
            
            if ii == 2:
                gpt("set format x")
                gpt.xlabel("Time [s]")
            
            gpt.plot(points, fit)


def str2orbit(line):
    line_split = line.split()
    
    return [float(line_split[0]), float(line_split[1]), float(line_split[2])]

File: src/aux/satorbit.cc
/* Copyright (C) 2018  István Bozsó
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <cmath>

#include "satorbit.hh"

static inline double norm(cdouble x, cdouble y, cdouble z)
{
    return sqrt(x * x + y * y + z * z);
}

// Calculate satellite position based on fitted polynomial orbits at time
static inline void calc_pos(const poly_fit& orb, double time, cart& pos)
{
    uint n_poly = orb.deg + 1, is_centered = orb.is_centered;
    double x = 0.0, y = 0.0, z = 0.0;
    
    const double *coeffs = orb.coeffs, *mean_coords = orb.mean_coords;
    
    if (is_centered)
        time -= orb.mean_t;
    
    if(n_poly == 2) {
        x = coeffs[0] * time + coeffs[1];
        y = coeffs[2] * time + coeffs[3];
        z = coeffs[4] * time + coeffs[5];
    }
    else {
        x = coeffs[0]           * time;
        y = coeffs[n_poly]      * time;
        z = coeffs[2 * n_poly]  * time;

        FOR(ii, 1, n_poly - 1) {
            x = (x + coeffs[             ii]) * time;
            y = (y + coeffs[    n_poly + ii]) * time;
            z = (z + coeffs[2 * n_poly + ii]) * time;
        }
        
        x += coeffs[    n_poly - 1];
        y += coeffs[2 * n_poly - 1];
        z += coeffs[3 * n_poly - 1];
    }
    
    if (is_centered) {
        x += mean_coords[0];
        y += mean_coords[1];
        z += mean_coords[2];
    }
    
    pos.x = x; pos.y = y; pos.z = z;
} // calc_pos

static inline double dot_product(const poly_fit& orb, cdouble X, cdouble Y,
                                 cdouble Z, double time)
{
    /* Calculate dot product between satellite velocity vector and
     * and vector between ground position and satellite position. */
    
    double dx, dy, dz, sat_x = 0.0, sat_y = 0.0, sat_z = 0.0,
                       vel_x, vel_y, vel_z, power, inorm;
    uint n_poly = orb.deg + 1;
    
    const double *coeffs = orb.coeffs, *mean_coords = orb.mean_coords;
    
    if (orb.is_centered)
        time -= orb.mean_t;
    
    // linear case 
    if(n_poly == 2) {
        sat_x = coeffs[0] * time + coeffs[1];
        sat_y = coeffs[2] * time + coeffs[3];
        sat_z = coeffs[4] * time + coeffs[5];
        
        vel_x = coeffs[0]; vel_y = coeffs[2]; vel_z = coeffs[4];
    }
    // evaluation of polynom with Horner's method
    else {
        
        sat_x = coeffs[0]           * time;
        sat_y = coeffs[n_poly]      * time;
        sat_z = coeffs[2 * n_poly]  * time;

        FOR(ii, 1, n_poly - 1) {
            sat_x = (sat_x + coeffs[             ii]) * time;
            sat_y = (sat_y + coeffs[    n_poly + ii]) * time;
            sat_z = (sat_z + coeffs[2 * n_poly + ii]) * time;
        }
        
        sat_x += coeffs[    n_poly - 1];
        sat_y += coeffs[2 * n_poly - 1];
        sat_z += coeffs[3 * n_poly - 1];
        
        vel_x = coeffs[    n_poly - 2];
        vel_y = coeffs[2 * n_poly - 2];
        vel_z = coeffs[3 * n_poly - 2];
        
        FOR(ii, 0, n_poly - 3) {
            power = double(n_poly - 1.0 - ii);
            vel_x += ii * coeffs[             ii] * pow(time, power);
            vel_y += ii * coeffs[    n_poly + ii] * pow(time, power);
            vel_z += ii * coeffs[2 * n_poly + ii] * pow(time, power);
        }
    }
    
    if (orb.is_centered) {
        sat_x += mean_coords[0];
        sat_y += mean_coords[1];
        sat_z += mean_coords[2];
    }
    
    // satellite coordinates - GNSS coordinates
    dx = sat_x - X;
    dy = sat_y - Y;
    dz = sat_z - Z;
    
    // product of inverse norms
    inorm = (1.0 / norm(dx, dy, dz)) * (1.0 / norm(vel_x, vel_y, vel_z));
    
    return (vel_x * dx  + vel_y * dy  + vel_z * dz) * inorm;
} // dot_product

// Compute the sat position using closest approche.
static inline void closest_appr(const poly_fit& orb, cdouble X, cdouble Y,
                                cdouble Z, cuint max_iter, cart& sat_pos)
{
    // first, last and middle time, extending the time window by 5 seconds
    double t_start = orb.start_t - 5.0,
           t_stop  = orb.stop_t + 5.0,
           t_middle = 0.0;
    
    // dot products
    double dot_start, dot_middle = 1.0;

    // iteration counter
    uint itr = 0;
    
    dot_start = dot_product(orb, X, Y, Z, t_start);
    
    while( fabs(dot_middle) > 1.0e-11 && itr < max_iter) {
        t_middle = (t_start + t_stop) / 2.0;

        dot_middle = dot_product(orb, X, Y, Z, t_middle);
        
        // change start for middle
        if ((dot_start * dot_middle) > 0.0) {
            t_start = t_middle;
            dot_start = dot_middle;
        }
        // change  end  for middle
        else
            t_stop = t_middle;

        itr++;
    } // while
    
    // calculate satellite position at middle time
    calc_pos(orb, t_middle, sat_pos);
} // closest_appr

void ell_cart (cdouble lon, cdouble lat, cdouble h,
               double& x, double& y, double& z)
{
    double n = WA / sqrt(1.0 - E2 * sin(lat) * sin(lat));

    x = (              n + h) * cos(lat) * cos(lon);
    y = (              n + h) * cos(lat) * sin(lon);
    z = ( (1.0 - E2) * n + h) * sin(lat);

} // ell_cart

void cart_ell(cdouble x, cdouble y, cdouble z,
              double& lon, double& lat, double& h)
{
    double n, p, o, so, co;

    n = (WA * WA - WB * WB);
    p = sqrt(x * x + y * y);

    o = atan(WA / p / WB * z);
    so = sin(o); co = cos(o);
    o = atan( (z + n / WB * so * so * so) / (p - n / WA * co * co * co) );
    so = sin(o); co = cos(o);
    n = WA * WA / sqrt(WA * co * co * WA + WB * so * so * WB);

    lat = o;
    
    o = atan(y/x); if(x < 0.0) o += M_PI;
    lon = o;
    h = p / co - n;
} // cart_ell


void calc_azi_inc(const poly_fit& orb, cdouble X, cdouble Y,
                  cdouble Z, cdouble lon, cdouble lat,
                  cuint max_iter, double& azi, double& inc)
{
    double xf, yf, zf, xl, yl, zl, t0;
    cart sat;
    
    // satellite closest approache cooridantes
    closest_appr(orb, X, Y, Z, max_iter, sat);
    
    xf = sat.x - X;
    yf = sat.y - Y;
    zf = sat.z - Z;
    
    // estiamtion of azimuth and inclination
    xl = - sin(lat) * cos(lon) * xf
         - sin(lat) * sin(lon) * yf + cos(lat) * zf ;
    
    yl = - sin(lon) * xf + cos(lon) * yf;
    
    zl = + cos(lat) * cos(lon) * xf
         + cos(lat) * sin(lon) * yf + sin(lat) * zf ;
    
    t0 = norm(xl, yl, zl);
    
    inc = acos(zl / t0) * RAD2DEG;
    
    if(xl == 0.0) xl = 0.000000001;
    
    double temp_azi = atan(fabs(yl / xl));
    
    if( (xl < 0.0) && (yl > 0.0) ) temp_azi = M_PI - temp_azi;
    if( (xl < 0.0) && (yl < 0.0) ) temp_azi = M_PI + temp_azi;
    if( (xl > 0.0) && (yl < 0.0) ) temp_azi = 2.0 * M_PI - temp_azi;
    
    temp_azi *= RAD2DEG;
    
    if(temp_azi > 180.0)
        temp_azi -= 180.0;
    else
        temp_azi += 180.0;
    
    azi = temp_azi;
} // calc_azi_inc
